moquette-0.16/broker/src/main/java/io/moquette/broker/config/ConfigurationParser.java

```diff
@@ -0,0 +1,119 @@
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.IOException;
+import java.io.Reader;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.text.ParseException;
+import java.util.Properties;
+
+/**
+ * Mosquitto configuration parser.
+ *
+ * A line that at the very first has # is a comment Each line has key value format, where the
+ * separator used it the space.
+ */
+class ConfigurationParser {
+
+    private static final Logger LOG = LoggerFactory.getLogger(ConfigurationParser.class);
+
+    private Properties m_properties = new Properties();
+
+    /**
+     * Parse the configuration from file.
+     */
+    void parse(File file) throws ParseException {
+        if (file == null) {
+            LOG.warn("parsing NULL file, so fallback on default configuration!");
+            return;
+        }
+        if (!file.exists()) {
+            LOG.warn(
+                    String.format(
+                            "parsing not existing file %s, so fallback on default configuration!",
+                            file.getAbsolutePath()));
+            return;
+        }
+        try {
+            Reader reader = Files.newBufferedReader(file.toPath(), StandardCharsets.UTF_8);
+            parse(reader);
+        } catch (IOException fex) {
+            LOG.warn("parsing not existing file {}, fallback on default configuration!", file.getAbsolutePath(), fex);
+        }
+    }
+
+    /**
+     * Parse the configuration
+     *
+     * @throws ParseException
+     *             if the format is not compliant.
+     */
+    void parse(Reader reader) throws ParseException {
+        if (reader == null) {
+            // just log and return default properties
+            LOG.warn("parsing NULL reader, so fallback on default configuration!");
+            return;
+        }
+
+        BufferedReader br = new BufferedReader(reader);
+        String line;
+        try {
+            while ((line = br.readLine()) != null) {
+                int commentMarker = line.indexOf('#');
+                if (commentMarker != -1) {
+                    if (commentMarker == 0) {
+                        // skip its a comment
+                        continue;
+                    } else {
+                        // it's a malformed comment
+                        throw new ParseException(line, commentMarker);
+                    }
+                } else {
+                    if (line.isEmpty() || line.matches("^\\s*$")) {
+                        // skip it's a black line
+                        continue;
+                    }
+
+                    // split till the first space
+                    int delimiterIdx = line.indexOf(' ');
+                    String key = line.substring(0, delimiterIdx).trim();
+                    String value = line.substring(delimiterIdx).trim();
+
+                    m_properties.put(key, value);
+                }
+            }
+        } catch (IOException ex) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)
Similar issue at line numbers 110.